### PR TITLE
Setup Vaultwarden on Heztner Cloud (securely)

### PR DESCRIPTION
**Pull request by mhupfauer**
Hi *,

I'm aware that this tutorial already exists but I think it can be improved significantly in terms of security which is imho the most important thing here.

I have read and understood the Contributor's Certificate of Origin available at the end of
https://raw.githubusercontent.com/hetzneronline/community-content/master/tutorial-template.md
and I hereby certify that I meet the contribution criteria described in it.
Signed-off-by: Markus Hupfauer [markus@hupfauer.one](mailto:markus@hupfauer.one)

### CODE REVIEW

File: tutorials/how-to-set-up-vaultwarden/01.en.md [hunk 1]
```diff
@@ -2,269 +2,420 @@
...
+1. `sender name` <- Replace with Vaultwarden or something to instantly tell this e-mail is from the password server
 
-![Vault](images/Vault.png "BitWarden Web Vault with passwords")
+1. `smtp username` <- In most cases your e-mail address except on AWS SES where you would need to create extra IAM users
 
-### Disable Anonymous User Sign Up
+1. `smtp-password` <- Your smtp password
 
-You can also choose to disable user registration as an extra security measure. To prevent abuse of the service, this is advised on servers that are publicly accessible.
+1. `Europe/Berlin` <- If you dont live in the GTM+1 Timezone change this
```
Comment (svenja11) [inline tutorials/how-to-set-up-vaultwarden/01.en.md:232]:
```suggestion
1. `Europe/Berlin` <- If you don't live in the GTM+1 Timezone change this
```

File: tutorials/how-to-set-up-vaultwarden/01.en.md [hunk 2]
```diff
@@ -2,269 +2,420 @@
...
+
+## Step 1 - Setup new Server
+
+Do not host Vaultwarden and other applications on the same host if not absolutely required. 
+The attack surface should be as small as possible and any added complexity from other applications will introduces unnecessary uncertainty.
+
+1. Open Hetzner Cloud Console
+1. Click "New Project +" and create a new project
+1. Click "Add Server" and select a datacenter which is closest to your users
+1. Select Debian 11 as OS
```
Comment (svenja11) [inline tutorials/how-to-set-up-vaultwarden/01.en.md:52]: It says to select "Debian 11". However, in the code below, you are using the Docker repository for Ubuntu. This doesn't work. See:
```
source: deb [arch=amd64] https://download.docker.com/linux/ubuntu $RELEASE stable
```

File: tutorials/how-to-set-up-vaultwarden/01.en.md [hunk 3]
```diff
@@ -2,269 +2,420 @@
...
+1. Click "Add ssh key" and paste your ssh-key. On Windows 10 you can get it via powershell `cat $env:USERPROFILE\.ssh\id_rsa.pub`
+1. Name the server appropriately and click "Create & Buy now"
+
+```yml
+#cloud-config
+package_upgrade: true
+apt:
+  sources:
+    docker.list:
+      source: deb [arch=amd64] https://download.docker.com/linux/ubuntu $RELEASE stable
```
Comment (svenja11) [inline tutorials/how-to-set-up-vaultwarden/01.en.md:64]: See my comment about "Create a server". There, Debian 11 is used as OS but this repository is for Ubuntu.

**Review by svenja11 — COMMENTED** [review]
Thank you for your contribution @mhupfauer. I read through your tutorial and I added some comments. Most of them are just suggestions. Let me know if you have any questions.

**Comment by blinkeye** [comment]
Thank you for improving the documentation. I followed the steps but this is not working for me:

1. I don't get an SSL certificate, I can ignore/accept the self-signed certificate and use configure my setup though
2. if I enable 2FA I can't login anymore and eventually receive a mail with `Incomplete Two-Step Login`

I compared the API calls with the official Bitwarden web client and I see that both issue:

1. `identity/accounts/prelogin`
2. followed by a HTTP 400 `identity/connect/token`

Bitwarden returns:

```bash
{"error":"invalid_grant","error_description":"Two factor required.","TwoFactorProviders":["0","2"], ...
```
while vaultwarden with this setup here returns just:

```bash
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>400 Bad Request</title>
</head><body>
<h1>Bad Request</h1>
<p>Your browser sent a request that this server could not understand.<br />
</p>
</body></html>
```
and stops, I can't enter anything. I tried both Duo and an Authenticator app, same behavior.

I see that the `vaultwarden` container logs:

```bash
vaultwarden  | [2023-02-05 18:22:26.453][error][ERROR] 2FA token not provided
```

whereas the waf logs the 400 bad request.

**Comment by svenja11** [comment]
@mhupfauer Do you still plan to work on this tutorial?

**Pull request merged**